feat(router): update router to support query params in intent links

pedrobonamin opened the pull request:
### Description
As part of releases we are introducing a query param that needs to be persisted in the url, `perspective`

It will determine which `perspective` the user has selected and we need to use the `IntentLink` component to change the perspective the user has selected when navigating to a version document.

How it will be used:
```ts
          <IntentLink
            {...linkProps}
            intent="edit"
            params={{
              id: getPublishedId(documentId, true),
              type: documentTypeName,
            }}
            searchParams={[['perspective', `bundle.${release.name}`]]}
            ref={ref}
          />
```
Path example: `/test/intent/edit/id=c0c99e43-40de-4d8b-abec-a710d5752ea2;type=book/?perspective=bundle.pedro-summer` 

see https://github.com/sanity-io/sanity/pull/7092/files#diff-729068b8f99a9501225501d3379505aa1a85162b6e41aa9654b7c5aa8a83d067R122-R139 for further details.

<!--
What changes are introduced?
Why are these changes introduced?
What issue(s) does this solve? (with link, if possible)
-->

### What to review

Is this change correct? Is there other way to handle this?

<!--
What steps should the reviewer take in order to review?
What parts/flows of the application/packages/tooling is affected?
-->

### Testing

Added tests for `IntentLink` and for `resolveIntentState` function

<!--
Did you add sufficient testing for this change?
If not, please explain how you tested this change and why it was not
possible/practical for writing an automated test
-->

### Notes for release

Adds support in intent links for `_searchParams: SearchParam[]`. 

<!--
A description of the change(s) that should be used in the release notes.
If this is PR is a partial implementation of a feature and is not enabled by default, please
call this out explicitly here so that it does not get included in the release notes.
-->

binoy14 commented on packages/sanity/src/core/studio/router/helpers.ts:
```diff
@@ -125,6 +125,7 @@ export function resolveIntentState(
 
     const nextUrlState: Record<string, unknown> = {
       ...prevState,
+      _searchParams: nextState._searchParams,
```
Wondering, what's the reason to use `_searchParams` instead of `searchParams`

pedrobonamin replied:
This is because the router expects the search params to be in a `_searchParams` property.
I've added an update to the `RouterEvent` interface to reflect that the state returned from this function should be a RouterState.
```diff
export interface RouterStateEvent {
  type: 'state'
+ state: RouterState
- state: Record<string, unknown> 
  isNotFound: boolean
}
```

RouterState is the following:
`export type RouterState = Record<string, unknown> & {_searchParams?: SearchParam[]}
`

pedrobonamin replied:
https://github.com/sanity-io/sanity/pull/7095/commits/646f433231a5f0ca8f0fd9baacc195e539c4257e

binoy14 replied:
understood, thanks for confirming

binoy14 reviewed (commented):
Couple questions

1. Is there a reason there would not be a need for release notes for this change?
2. Is it possible to add tests for this change? Not sure if there are existing tests for Router that you can piggy back on. 

Looks good otherwise besides the comments

pedrobonamin commented:
> Couple questions
> 
> 1. Is there a reason there would not be a need for release notes for this change?
> 2. Is it possible to add tests for this change? Not sure if there are existing tests for Router that you can piggy back on.
> 
> Looks good otherwise besides the comments

Thanks for the review @binoy14 .

Added test for the change in the IntentLink component, that now supports query params and added test for the resolveIntentState function that it now supports query params

binoy14 reviewed (approved):
Makes sense to me, thanks!
